Rubberband: switch between v2 and v3 at runtime

ronso0 opened the pull request:
Based on @Swiftb0y's #4853

Ready for review, though draft since I'll rebase if necessary / when #4853 is merged.

ronso0 committed:
RubberBand: allow switching v2/v3 at runtime

Swiftb0y commented:
Thanks for the commit. It doesn't build unfortunately (fix is trivial). I'll finish #4853 and and then review this. It'll be a couple of days though.

daschuer commented on src/engine/bufferscalers/enginebufferscalerubberband.cpp:
```diff
@@ -17,16 +17,21 @@ using RubberBand::RubberBandStretcher;
 // This is the default increment from RubberBand 1.8.1.
 size_t kRubberBandBlockSize = 256;
 
+#define RUBBERBANDV3 (RUBBERBAND_API_MINOR_VERSION >= 2 && RUBBERBAND_API_MINOR_VERSION >= 7)
```
```suggestion
#define RUBBERBANDV3 (RUBBERBAND_API_MINOR_VERSION >= 2 && RUBBERBAND_API_MINOR_VERSION >= 7)
```
```suggestion
#define RUBBERBANDV3 (RUBBERBAND_API_MAJOR_VERSION >= 2 && RUBBERBAND_API_MINOR_VERSION >= 7)
```

daschuer commented on src/engine/bufferscalers/enginebufferscalerubberband.cpp:
```diff
@@ -109,10 +117,18 @@ void EngineBufferScaleRubberBand::onSampleRateChanged() {
     }
+    RubberBandStretcher::Options rubberbandOptions =
+            RubberBandStretcher::OptionProcessRealTime;
+    if (m_useEngineV3) {
+        rubberbandOptions |= RubberBandStretcher::OptionEngineFiner;
+    }
```
```suggestion
    if (m_useEngineV3) {
        rubberbandOptions |= RubberBandStretcher::OptionEngineFiner;
    }
```
```suggestion
#if RUBBERBANDV3
    if (m_useEngineV3) {
        rubberbandOptions |= RubberBandStretcher::OptionEngineFiner;
    }
#endif     
```

daschuer commented on src/engine/bufferscalers/enginebufferscalerubberband.h:
```diff
@@ -45,4 +50,6 @@ class EngineBufferScaleRubberBand : public EngineBufferScale {
     // Holds the playback direction
     bool m_bBackwards;
+
+    bool m_useEngineV3;
```
The term engine V3 becomes weak, if they introduce more options in the lifetime of V3
```suggestion
    bool m_useEngineFiner;
```

daschuer commented on src/engine/enginebuffer.cpp:
```diff
@@ -62,6 +62,8 @@ constexpr SINT kSamplesPerFrame = 2; // Engine buffer uses Stereo frames only
 // Rate at which the playpos slider is updated
 constexpr int kPlaypositionUpdateRate = 15; // updates per second
 
+#define RUBBERBANDV3 (RUBBERBAND_API_MINOR_VERSION >= 2 && RUBBERBAND_API_MINOR_VERSION >= 7)
```
```suggestion
#define RUBBERBANDV3 (RUBBERBAND_API_MINOR_VERSION >= 2 && RUBBERBAND_API_MINOR_VERSION >= 7)
```

daschuer replied:
But it would be nice to hide the dependency here.

ronso0 replied:
I don't spot the difference in your suggestion?

Swiftb0y replied:
Agree, including the entirety of RubberbandStretcher.h just for `RUBBERBAND_API_MAJOR-/MINOR_VERSION` is quite a big dependency, but what other choice do we have?

ronso0 replied:
would a global const defined in enginebufferscalerubberband.h work?

Swiftb0y replied:
Not really... We need to differentiate at build time and the only way is by including the header. It would be nice if rubberband had a separate header that just includes the version, but it doesn't.

daschuer replied:
My original point was that the EngineScaler should handle the dependency not EngineBuffer. 
Than EngieBuffer can use a member of EngineScaler to query about that.

ronso0 replied:
understood. 
> We need to differentiate at build time

That as well, thus my naive question about global const (set at compile time in the EngineScaler).

> Than EngieBuffer can use a member of EngineScaler to query about that.

and that, too. Though how is that supossed to work for the enum in the EngineBuffer header?

Swiftb0y replied:
> That as well, thus my naive question about global const (set at compile time in the EngineScaler).

The issue with globals consts is that they get to be evaluated after the preprocessor finished. Since we want to do conditional compilation using `#if`, we need this value sooner, so we need a `#define` for that then. `RubberBandStretcher.h` is a pretty lean header (90% is just comments) so I guess its not a big problem to include that a bunch of times.

daschuer replied:
> Though how is that supossed to work for the enum in the EngineBuffer header?

My idea was to remove all the guards from the static code as if we are at a V3 build. If the engine Finer is not available the option need to be hidden from the GUI, that's all.

ronso0 replied:
Sure, but hiding it in the GUI seems more complicated than working with the guards in both EngineBuffer and EngineBufferScaleRubberBand.

Are you suggesting to first populate the keylock combobox with _all_ values, then (probably in 
[DlgPrefSound::loadSettings](https://github.com/mixxxdj/mixxx/blob/e1ee93311268101f06a761a4e258aa3b73a76a73/src/preferences/dialog/dlgprefsound.cpp#L499-L502)) somehow query EngineBuffer / EngineBufferScaleRubberBand to decide if we remove `RubberBand finer` from the combobox (and also reset the config to RubberBand (faster)) ?

daschuer commented on src/engine/enginebuffer.cpp:
```diff
@@ -802,13 +800,29 @@ void EngineBuffer::slotKeylockEngineChanged(double dIndex) {
+    default:
+#ifdef RUBBERBANDV3
+        // trigger reconstruction of RubberBandStretcher with v2 options
+        m_pScaleRB->useEngineVersion3(false);
```
```suggestion
        m_pScaleRB->useEngineFiner(false);
```

daschuer commented on src/engine/enginebuffer.h:
```diff
@@ -77,6 +78,9 @@ class EngineBuffer : public EngineObject {
     enum KeylockEngine {
         SOUNDTOUCH,
         RUBBERBAND,
+#if (RUBBERBAND_API_MINOR_VERSION >= 2 && RUBBERBAND_API_MINOR_VERSION >= 7)
```
I think we can remove this guard here

ronso0 replied:
nope, the Keylock Engine combobox is populated with those values.

daschuer replied:
Yes, but defining the enums value in any case will work.

daschuer commented on src/engine/enginebuffer.h:
```diff
@@ -148,6 +152,10 @@ class EngineBuffer : public EngineObject {
             return tr("Soundtouch (faster)");
         case RUBBERBAND:
```
```suggestion
        case RUBBERBAND_FASTER:
```

daschuer commented on src/engine/enginebuffer.h:
```diff
@@ -148,6 +152,10 @@ class EngineBuffer : public EngineObject {
         case RUBBERBAND:
             return tr("Rubberband (better)");
+#if (RUBBERBAND_API_MINOR_VERSION >= 2 && RUBBERBAND_API_MINOR_VERSION >= 7)
+        case RUBBERBAND_R3:
```
```suggestion
        case RUBBERBAND_FINER:
```

daschuer commented on src/engine/enginebuffer.h:
```diff
@@ -148,6 +152,10 @@ class EngineBuffer : public EngineObject {
             return tr("Rubberband (better)");
+#if (RUBBERBAND_API_MINOR_VERSION >= 2 && RUBBERBAND_API_MINOR_VERSION >= 7)
+        case RUBBERBAND_R3:
+            return tr("Rubberband R3 (even better)");
```
```suggestion
            return tr("Rubberband R3 (near-hi-fi quality)");
```
taken from https://breakfastquay.com/rubberband/why.html

daschuer commented on src/engine/bufferscalers/enginebufferscalerubberband.h:
```diff
@@ -17,6 +17,9 @@ class EngineBufferScaleRubberBand : public EngineBufferScale {
     ~EngineBufferScaleRubberBand() override;
 
+    // Enable engine v3 if available
+    void useEngineVersion3(bool enable);
```
We may add a `hasEngineFiner()` to get around the guards in EngineBuffer

daschuer reviewed (changes requested):
Thank you. I have left some comments for improvement.

ronso0 committed:
EngineBuffer: remove RubberBand include and version guards, Note:

now the KeylockEngine enum always contains RUBBERBAND_FINER, thus the selector
in Preferences > Sound Hardware will always show it, too.

ronso0 commented:
Well, this now builds and I fixed the test (adjusted to changed enum).
I got around the guards in EngineBuffer, but as I said, I don't see a way to hide `RubberBand finer)` from the pref combobox.¹
I guess with a little work it can at least switch back to RubberBand (faster) if (finer) is not available, but that's bad UX iMO

¹of course there are ways, like using a CO/ControlProxy "isRubberBandV3Available" but that's ugly and hacky (even for my taste) and error-prone.

ronso0 committed:
Pref Sound: reset to RubberBand v2 if v3 isn't available

daschuer commented on src/preferences/dialog/dlgprefsound.cpp:
```diff
@@ -496,7 +505,7 @@ void DlgPrefSound::loadSettings(const SoundManagerConfig &config) {
     }
 
-    // Default keylock is Rubberband.
+    // Default keylock is Rubberband v2
```
```suggestion
    // Default keylock is Rubberband Faster
```

daschuer commented:
Idea:
How about using: 
EngineBufferScaleRubberBand::isEngineFinerAvailable()
And populate the "Finer" field only if it is available or if it was selected before. 
If we have a Rubberband 2 build we may change pevefix the text with "not installed" or such.

Swiftb0y commented:
Not a fan of hiding options for no apparent reason. How much more effort would it be to grey it out instead and provide a tooltip explaining why its not available?

daschuer commented:
Since we do not dynamically link to V2 and V3 users with V2 have no chance to "ungray" the grayed out option. That's why I have proposed to hide it. If we want to use it as a teaser, we need at least a hint how to ungray it. 

The only case where wrongly selected is, if the user has selected it in an experimental build and starts Mixxx without V3.  Currently the combo-box is empty, but it uses Rubberband Fast. Here we can actually switch back to Rubberband Fast, or inform the user that the selection is not available. This is really an edge case, not worth to code to much around it. IMHO 

What do you think?

Swiftb0y commented:
> That's why I have proposed to hide it. If we want to use it as a teaser, we need at least a hint how to ungray it.

Sure, thats what I would've put in tooltip for the greyed out option.

ronso0 commented:
> Idea: How about using: EngineBufferScaleRubberBand::isEngineFinerAvailable() And populate the "Finer" field only if it is available or if it was selected before. If we have a Rubberband 2 build we may change pevefix the text with "not installed" or such.

Again: if you have any solution how to do this _without_ hazzle nad hacks, please present it.
IMO no solution is easier and less error-prone than simply using the guards for the enum in enginebuffer. As @Swiftb0y pointed out the rubberband header is actually rather small.

daschuer commented:
> I guess with a little work it can at least switch back to RubberBand (faster) if (finer) is not available, but that's bad UX iMO

In a test branch, I have tried to keep the bad Rubberband value, but since it switches automatically back to default, the value does not persist after another restart. I think the use case is, using Finer in the recent R3 build, test an older Mixxx and go back to the R3 build without loosing the selection of the Finer engine in preference. 

With 2.3, the combobox is empty in this case. I think we should display "Unknown (bad value)" for main R2 builds instead and hide the value if a valid value was selected.

daschuer committed:
Keep Rubberband Finer selected during testing a Rubberband V2 build

Co-authored-by: Swiftb0y <12380386+Swiftb0y@users.noreply.github.com>
Co-authored-by: ronso0 <ronso0@mixxx.org>

daschuer commented:
Here it is: https://github.com/ronso0/mixxx/pull/26

It displays now "Unknown, using Rubberband (better)" in case of R3 was selected but is not available in the running build.

daschuer committed:
Introduce EngineBuffer::defaultKeylockEngine() for a single source of the default

daschuer committed:
Add a comment about enum KeylockEngine and backward compatibility

ywwg commented:
Is the rubberband library version a compile-time selection?  If so, I would vote to hide the option in the UI.  Having a grayed-out option in the prefs for a value which literally cannot be enabled by the user does not communicate anything useful.

Other examples of the same approach:
* on linux, I am not presented with ASIO as a sound card access option because there's no way I could possibly use it.  
* If vinyl control is disabled at compile time, relevant preferences are removed.
* We do not show effects that are not installed on the system.

ywwg commented:
(is this PR still a draft?)

ronso0 commented:
@ywwg unavailable engines are hidden in ronso0#26, please take a look.
Still draft until I merge that and clean up this PR, which I'll probably find time for in the next 2-5 days.

ronso0 commented:
With commits from https://github.com/ronso0/mixxx/pull/26 this is now ready to roll!

Swiftb0y reviewed (changes requested):
I didn't want to put more work on you so I went ahead and implemented all my review suggestions for you: ronso0/mixxx/pull/27

Swiftb0y committed:
refactor(enginebuffer): Make EngineBuffer::KeylockEngine an enum class (#27)

refactor(prefs): improve keylockengine-related typesafety

Swiftb0y commented on src/preferences/dialog/dlgprefsound.cpp:
```diff
@@ -681,8 +691,14 @@ void DlgPrefSound::slotResetToDefaults() {
-    keylockComboBox->setCurrentIndex(EngineBuffer::RUBBERBAND);
-    m_pKeylockEngine->set(EngineBuffer::RUBBERBAND);
+
+    int keylock_engine = static_cast<int>(EngineBuffer::defaultKeylockEngine());
+    int index = keylockComboBox->findData(keylock_engine);
+    DEBUG_ASSERT(index >= 0);
+    if (index >= 0) {
+        keylockComboBox->setCurrentIndex(index);
+    }
+    m_pKeylockEngine->set(keylock_engine);
```
I forgot to fix this as well. My fault. 

```suggestion
    const auto keylockEngine = EngineBuffer::defaultKeylockEngine();
    const int index = keylockComboBox->findData(QVariant::fromValue(keyLockEngine));
    DEBUG_ASSERT(index >= 0);
    if (index >= 0) {
        keylockComboBox->setCurrentIndex(index);
    }
    m_pKeylockEngine->set(static_cast<double>(keyLockEngine));
```

Swiftb0y reviewed (commented):
lgtm otherwise

ronso0 committed:
keylock engine: use typesafe values also in DlgPrefSound::slotResetToDefaults

Co-authored-by: Swiftb0y <12380386+Swiftb0y@users.noreply.github.com>

ronso0 commented:
would be nice to have two :heavy_check_mark: before I hit merge : )

daschuer commented:
I can do a final test after the build issues are solved.

Swiftb0y commented:
Sorry. The suggestion I posted contained capitalization typos. Sent you a PR.

ronso0 commented:
I don't understand what's wrong with pre-commit
https://github.com/mixxxdj/mixxx/runs/7679981732?check_suite_focus=true
`
WARNING: The directory '/github/home/.cache/pip' or its parent directory is not owned or is not writable by the current user. The cache has been disabled. Check the permissions and owner of that directory. If executing pip with sudo, you may want sudo's -H flag.
`

Swiftb0y commented:
I think versioning issues. seems to me like `distlib` is installed as part of the systems python installation but something wants a newer distlib version. So in order to install that, it has to uninstall the older version which is part of system install so it can't remove that. I'm not sure. I'm investigating.

Swiftb0y commented:
@daschuer pre-commit issues are unrelated. Can you do your final test before merge?

daschuer commented:
Just did a brief test and all works as expected. Thank you all for the good collaboration.

daschuer commented:
@Swiftb0y merge? Your review is still red.

Swiftb0y reviewed (approved):
whoops, lgtm.

daschuer commented:
Thank you.
